kbs/docs/config.md
# KBS Configuration File

The Confidential Containers KBS properties can be configured through a
TOML-formatted configuration file.

> NOTE: Additional formats such as YAML and JSON are supported. Other formats
> supported by the `config` crate may be supported as well. This document uses
> TOML in the configuration examples.

The location of the configuration file is passed to the KBS binary using the
`-c` or `--config-file` command line option, or using the `KBS_CONFIG_FILE`
environment variable.

## Configurable Properties

The following sections list the KBS properties which can be set through the configuration file.

### HTTP Server Configuration

The following properties can be set under the `[http_server]` section.

| Property               | Type         | Description                                      |  Required | Default                  |
|------------------------|--------------|--------------------------------------------------|----------|--------------------------|
| `sockets`              | String array | One or more sockets to listen on.                | No       | `["127.0.0.1:8080"]`     |
| `insecure_http`        | Boolean      | Don't use TLS for the KBS HTTP endpoint.         | No       | `false`                  |
| `private_key`          | String       | Path to a private key file to be used for HTTPS. | No       | None                     |
| `certificate`          | String       | Path to a certificate file to be used for HTTPS. | No       | None                     |
| `payload_request_size` | Integer      | Request payload size in mega bytes.              | No       | 2                        |
| `worker_count`         | Integer      | Number of HTTP actix worker threads              | No       | Num of logical CPU cores |

### Attestation Token Configuration

Attestation Token configuration controls attestation token verifications. This
is important when a resource retrievement is handled by KBS. Usually an attestation
token will be together with the request, and KBS will first verify the token.

The following properties can be set under the `[attestation_token]` section.

| Property              | Type         | Description                                                                                                                       | Default |
|-----------------------|--------------|-----------------------------------------------------------------------------------------------------------------------------------|---------|
| `trusted_jwk_sets`    | String Array | Valid Url (`file://` or `https://`) pointing to trusted JWKSets (local or OpenID) for Attestation Tokens trustworthy verification | Empty   |
| `trusted_certs_paths` | String Array | Trusted Certificates file (PEM format) for Attestation Tokens trustworthy verification                                            | Empty   |
| `extra_teekey_paths`  | String Array | User defined paths to the tee public key in the JWT body                                                                          | Empty   |
| `insecure_key`        | Boolean      | Whether to check the trustworthy of the JWK inside JWT. See comments.                                                             | `false` |

Each JWT contains a TEE Public Key. Users can use the `extra_teekey_paths` field to additionally specify the path of
this Key in the JWT.
Example of `extra_teekey_paths` is `/attester_runtime_data/tee-pubkey` which refers to the key
`attester_runtime_data.tee-pubkey` inside the JWT body claims. By default CoCo AS Token and Intel TA
Token TEE Public Key paths are supported.

For Attestation Services like CoCo-AS, the public key to verify the JWT will be given
in the token's `jwk` field (with or without the public key cert chain `x5c`).

- If `insecure_key` is set to `true`, KBS will ignore to verify the trustworthy of the `jwk`.
- If `insecure_key` is set to `false`, KBS will look up its `trusted_certs_paths` and the `x5c`
  field to verify the trustworthy of the `jwk`.

For Attestation Services like Intel TA, there will only be a `kid` field inside the JWT.
The `kid` field is used to look up the trusted jwk configured by KBS via `trusted_jwk_sets` to
verify the integrity and trustworthy of the JWT.

### Attestation Configuration

Attestation configuration defines the attestation service that KBS' RCAR protocol will leverage.

The following properties can be set under the `[attestation_service]` section.

Concrete attestation service can be set via `type` field. Supported attestation services are

- `coco_as_builtin`: CoCo AS that built inside KBS binary
- `coco_as_grpc`: CoCo AS service running remotely
- `intel_ta`: Intel&reg; Trust Authority

Due to different `type` field, properties are different.

#### Built-In CoCo AS

When `type` is set to `coco_as_builtin`, the following properties can be set.

> Built-In CoCo AS is available only when one or more of the following features are enabled:
> `coco-as-builtin`, `coco-as-builtin-no-verifier`

| Property                   | Type                        | Description                                              | Default                                                                                                       |
|----------------------------|-----------------------------|----------------------------------------------------------|---------------------------------------------------------------------------------------------------------------|
| `timeout`                  | Integer                     | The maximum time (in minutes) of the attestation session | 5                                                                                                             |
| `work_dir`                 | String                      | The location for Attestation Service to store data.      | First try from env `AS_WORK_DIR`. If no this env, then use `/opt/confidential-containers/attestation-service` |
| `policy_engine`            | String                      | Policy engine type. Valid values: `opa`                  | `opa`                                                                                                         |
| `rvps_config`              | [RVPSConfiguration][2]      | RVPS configuration                                       | See [RVPSConfiguration][2]                                                                                    |
| `attestation_token_broker` | [AttestationTokenBroker][1] | Attestation result token configuration.                  | See [AttestationTokenBroker][1]                                                                               |

[1]: #attestationtokenbroker

[2]: #rvps-configuration

##### AttestationTokenBroker

| Property         | Type                   | Description                                                                    | Required | Default                                                               |
|------------------|------------------------|--------------------------------------------------------------------------------|----------|-----------------------------------------------------------------------|
| `duration_min`   | Integer                | Duration of the attestation result token in minutes.                           | No       | `5`                                                                   |
| `issuer_name`    | String                 | Issure name of the attestation result token.                                   | No       | `CoCo-Attestation-Service`                                            |
| `developer_name` | String                 | The developer name to be used as part of the Verifier ID in the EAR            | No       | `https://confidentialcontainers.org`                                  |
| `build_name`     | String                 | The build name to be used as part of the Verifier ID in the EAR                | No       | Automatically generated from Cargo package and AS version             |
| `profile_name`   | String                 | The Profile that describes the EAR token                                       | No       | tag:github.com,2024:confidential-containers/Trustee`                  |
| `policy_dir`     | String                 | The path to the work directory that contains policies to provision the tokens. | No       | `/opt/confidential-containers/attestation-service/token/policies` |
| `signer`         | [TokenSignerConfig][1] | Signing material of the attestation result token.                              | No       | None                                                                  |

[1]: #tokensignerconfig

##### TokenSignerConfig

This section is **optional**. When omitted, an ephemeral RSA key pair is generated and used.

| Property    | Type   | Description                                              | Required |
|-------------|--------|----------------------------------------------------------|----------|
| `key_path`  | String | RSA Key Pair file (PEM format) path.                     | Yes      |
| `cert_url`  | String | RSA Public Key certificate chain (PEM format) URL.       | No       |
| `cert_path` | String | RSA Public Key certificate chain (PEM format) file path. | No       |

##### RVPS Configuration

| Property | Type   | Description                                                                                | Required | Default   |
|----------|--------|--------------------------------------------------------------------------------------------|----------|-----------|
| `type`   | String | It can be either `BuiltIn` (Built-In RVPS) or `GrpcRemote` (connect to a remote gRPC RVPS) | No       | `BuiltIn` |

##### BuiltIn RVPS

If `type` is set to `BuiltIn`, the following extra properties can be set

| Property  | Type                        | Description                                                                  | Required | Default   |
|-----------|-----------------------------|------------------------------------------------------------------------------|----------|-----------|
| `storage` | ReferenceValueStorageConfig | Configuration of the storage for reference values (`LocalFs` or `LocalJson`) | No       | `LocalFs` |

A `ReferenceValueStorageConfig` can either be of type `LocalFs` or `LocalJson`

For `LocalFs`, the following properties can be set

| Property    | Type   | Description                                        | Required | Default                                                             |
|-------------|--------|----------------------------------------------------|----------|---------------------------------------------------------------------|
| `file_path` | String | The path to the directory storing reference values | No       | `/opt/confidential-containers/attestation-service/reference_values` |

For `LocalJson`, the following properties can be set

| Property    | Type   | Description                                        | Required | Default                                                                  |
|-------------|--------|----------------------------------------------------|----------|--------------------------------------------------------------------------|
| `file_path` | String | The path to the file that storing reference values | No       | `/opt/confidential-containers/attestation-service/reference_values.json` |

##### Remote RVPS

If `type` is set to `GrpcRemote`, the following extra properties can be set

| Property  | Type   | Description                       | Required | Default           |
|-----------|--------|-----------------------------------|----------|-------------------|
| `address` | String | Remote address of the RVPS server | No       | `127.0.0.1:50003` |

#### gRPC CoCo AS

When `type` is set to `coco_as_grpc`, KBS will try to connect a remote CoCo AS for attestation. 
The following properties can be set.

> gRPC CoCo AS is available only when `coco-as-grpc` feature is enabled.

| Property    | Type    | Description                                                                                                                   | Default                  |
|-------------|---------|-------------------------------------------------------------------------------------------------------------------------------|--------------------------|
| `timeout`   | Integer | The maximum time (in minutes) between RCAR handshake's `auth` and `attest` requests                                           | 5                        |
| `as_addr`   | String  | The URL of the remote CoCoAS                                                                                                  | `http://127.0.0.1:50004` |
| `pool_size` | Integer | The connections between KBS and CoCoAS are maintained in a conenction pool. This property determines the max size of the pool | `100`                    |

#### Intel&reg; TA

When `type` is set to `intel_ta`, KBS will try to connect a remote Intel TA service for attestation. 
The following properties can be set.

> Intel Trust Authority AS is available only when the `intel-trust-authority-as` feature is enabled.

| Property                 | Type         | Description                                                                                            | Required | Default |
|--------------------------|--------------|--------------------------------------------------------------------------------------------------------|----------|---------|
| `timeout`                | Integer      | The maximum time (in minutes) between RCAR handshake's `auth` and `attest` requests                    | No       | 5       |
| `base_url`               | String       | Intel Trust Authority API URL.                                                                         | Yes      | -       |
| `api_key`                | String       | Intel Trust Authority API key.                                                                         | Yes      | -       |
| `certs_file`             | String       | URL to an Intel Trust Authority portal or path to JWKS file used for token verification.               | Yes      | -       |
| `policy_ids`             | String array | Quoted and comma-separated list of policy IDs defined in ITA portal.                                   | No       | `[]`    |
| `allow_unmatched_policy` | Boolean      | Whether policy matching is required. If no `policy_ids` are specified, policy matching is not checked. | No       | false   |

Detailed [documentation](https://docs.trustauthority.intel.com).

### Admin API Configuration

Multiple Admin backends are available. These control access to admin endpoints such as `set_policy`.
Today, the available backends are `DenyAll` (disables admin endpoints), `InsecureAllowAll` (for debugging)
and `Simple`.
By default, the simple backend will be used, but no personas will be enabled.
Use the `type` field to set the admin backend.

| Property          | Type    | Description                                  | Required | Default |
|-------------------|---------|----------------------------------------------|----------|---------|
| `type`            | String  | The backend used to validate admin requests. | No       | Simple  |

If the `Simple` backend is used, a list of admin personas can be provided, each with the following properties:

| Property          | Type   | Description                                                  | Required | Default |
|-------------------|--------|--------------------------------------------------------------|----------|---------|
| `id`              | String | A string used to identify the admin.                         | Yes      | Simple  |
| `public_key_path` | String | The path to the public key corresponding to the admin token. | Yes      | Simple  |
| `key_type`        | String | `EdDSA` for Ed25519 or `ES384` for P384                      | No       | EdDSA   |

### Policy Engine Configuration

The following properties can be set under the `[policy_engine]` section.

This section is **optional**. When omitted, a default configuration is used.

| Property      | Type   | Description                                                                                                | Required | Default                                        |

kbs/src/config.rs
// Copyright (c) 2023 by Alibaba.
// Licensed under the Apache License, Version 2.0, see LICENSE for details.
// SPDX-License-Identifier: Apache-2.0

use crate::admin::AdminConfig;
use crate::plugins::PluginsConfig;
use crate::policy_engine::PolicyEngineConfig;
use crate::token::AttestationTokenVerifierConfig;
use anyhow::anyhow;
use clap::Parser;
use config::{Config, File};
use serde::Deserialize;
use std::net::SocketAddr;
use std::path::{Path, PathBuf};

const DEFAULT_INSECURE_HTTP: bool = false;
const DEFAULT_SOCKET: &str = "127.0.0.1:8080";
const DEFAULT_PAYLOAD_REQUEST_SIZE: u32 = 2;

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(default)]
pub struct HttpServerConfig {
    /// Socket addresses (IP:port) to listen on, e.g. 127.0.0.1:8080.
    pub sockets: Vec<SocketAddr>,

    /// HTTPS private key.
    pub private_key: Option<PathBuf>,

    /// HTTPS Certificate.
    pub certificate: Option<PathBuf>,

    /// Insecure HTTP.
    /// WARNING: Using this option makes the HTTP connection insecure.
    pub insecure_http: bool,

    /// Request payload size in MB
    pub payload_request_size: u32,

    /// Number of worker threads for the actix-web server.
    /// If not specified, defaults to the number of logical CPU cores.
    pub worker_count: Option<usize>,
}

impl Default for HttpServerConfig {
    fn default() -> Self {
        Self {
            sockets: vec![DEFAULT_SOCKET.parse().expect("unexpected parse error")],
            private_key: None,
            certificate: None,
            insecure_http: DEFAULT_INSECURE_HTTP,
            payload_request_size: DEFAULT_PAYLOAD_REQUEST_SIZE,
            worker_count: None,
        }
    }
}

/// Contains all configurable KBS properties.
#[derive(Debug, Clone, Deserialize, PartialEq, Default)]
pub struct KbsConfig {
    /// Attestation token result broker config.
    #[serde(default)]
    pub attestation_token: AttestationTokenVerifierConfig,

    /// Configuration for the Attestation Service.
    #[cfg(feature = "as")]
    #[serde(default)]
    pub attestation_service: crate::attestation::config::AttestationConfig,

    /// Configuration for the KBS Http Server
    pub http_server: HttpServerConfig,

    /// Configuration for the KBS admin API
    pub admin: AdminConfig,

    /// Policy engine configuration used for evaluating whether the TCB status has access to
    /// specific resources.
    #[serde(default)]
    pub policy_engine: PolicyEngineConfig,

    #[serde(default)]
    pub plugins: Vec<PluginsConfig>,
}

impl TryFrom<&Path> for KbsConfig {
    type Error = anyhow::Error;

    /// Load `Config` from a configuration file. Supported formats are all formats supported by the
    /// `config` crate. See `KbsConfig` for schema information.
    fn try_from(config_path: &Path) -> Result<Self, Self::Error> {
        let c = Config::builder()
            .add_source(File::with_name(config_path.to_str().unwrap()))
            .build()?;

        c.try_deserialize()
            .map_err(|e| anyhow!("invalid config: {}", e))
    }
}

/// KBS command-line arguments.
#[derive(Debug, Parser)]
#[command(author, version, about, long_about = None)]
pub struct Cli {
    /// Path to a KBS config file. Supported formats: TOML, YAML, JSON and possibly other formats
    /// supported by the `config` crate.
    #[arg(short, long, env = "KBS_CONFIG_FILE")]
    pub config_file: String,
}

#[cfg(test)]
mod tests {
    use std::path::{Path, PathBuf};

    use crate::{
        admin::{
            simple::{SimpleAdminConfig, SimplePersonaConfig},
            AdminBackendType, AdminConfig,
        },
        config::{
            HttpServerConfig, DEFAULT_INSECURE_HTTP, DEFAULT_PAYLOAD_REQUEST_SIZE, DEFAULT_SOCKET,
        },
        plugins::{
            implementations::{
                resource::local_fs::LocalFsRepoDesc, RepositoryConfig, SampleConfig,
            },
            PluginsConfig,
        },
        policy_engine::{PolicyEngineConfig, DEFAULT_POLICY_PATH},
        token::AttestationTokenVerifierConfig,
    };

    use super::KbsConfig;

    #[cfg(feature = "coco-as-builtin")]
    use attestation_service::{
        ear_token::{EarTokenConfiguration, COCO_AS_ISSUER_NAME, DEFAULT_TOKEN_DURATION},
        rvps::{grpc::RvpsRemoteConfig, RvpsConfig, RvpsCrateConfig},
    };

    use reference_value_provider_service::storage::{local_fs, ReferenceValueStorageConfig};

    use rstest::rstest;

    #[rstest]
    #[case("test_data/configs/coco-as-grpc-1.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            trusted_certs_paths: vec!["/etc/ca".into(), "/etc/ca2".into()],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-grpc")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASGrpc(
                    crate::attestation::coco::grpc::GrpcConfig {
                        as_addr: "http://127.0.0.1:50001".into(),
                        pool_size: 100,
                    },
                ),
            timeout: 600,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: Some("/etc/kbs-private.key".into()),
            certificate: Some("/etc/kbs-cert.pem".into()),
            insecure_http: false,
            payload_request_size: DEFAULT_PAYLOAD_REQUEST_SIZE,
            worker_count: None,
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::DenyAll,
        },
        policy_engine: PolicyEngineConfig {
            policy_path: PathBuf::from("/etc/kbs-policy.rego"),
        },
        plugins: vec![PluginsConfig::Sample(SampleConfig {
            item: "value1".into(),
        }),
        PluginsConfig::ResourceStorage(RepositoryConfig::LocalFs(
            LocalFsRepoDesc {
                dir_path: "/tmp/kbs-resource".into(),
            },
        ))],
    })]
    #[case("test_data/configs/coco-as-builtin-1.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            trusted_certs_paths: vec![],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-builtin")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASBuiltIn(
                    attestation_service::config::Config {
                        work_dir: "/opt/coco/attestation-service".into(),
                        rvps_config: RvpsConfig::GrpcRemote(RvpsRemoteConfig {
                            address: "http://127.0.0.1:50003".into(),
                        }),
                        attestation_token_broker: EarTokenConfiguration {
                            duration_min: DEFAULT_TOKEN_DURATION,
                            issuer_name: COCO_AS_ISSUER_NAME.into(),
                            signer: None,
                            ..Default::default()
                        },
                        verifier_config: None,
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec![DEFAULT_SOCKET.parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: DEFAULT_INSECURE_HTTP,
            payload_request_size: DEFAULT_PAYLOAD_REQUEST_SIZE,
            worker_count: None,
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::DenyAll,
        },
        policy_engine: PolicyEngineConfig {
            policy_path: DEFAULT_POLICY_PATH.into(),
        },
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/intel-ta-1.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            trusted_jwk_sets: vec!["/etc/ca".into(), "/etc/ca2".into()],
            insecure_key: false,
            trusted_certs_paths: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "intel-trust-authority-as")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::IntelTA(
                    crate::attestation::intel_trust_authority::IntelTrustAuthorityConfig {
                        base_url: "example.io".into(),
                        api_key: "this-is-a-key".into(),
                        certs_file: "file:///etc/ita-cert.pem".into(),
                        allow_unmatched_policy: Some(true),
                        policy_ids: vec![],
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: Some("/etc/kbs-private.key".into()),
            certificate: Some("/etc/kbs-cert.pem".into()),
            insecure_http: false,
            payload_request_size: DEFAULT_PAYLOAD_REQUEST_SIZE,
            worker_count: None,
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::DenyAll,
        },
        policy_engine: PolicyEngineConfig {
            policy_path: PathBuf::from("/etc/kbs-policy.rego"),
        },
        plugins: vec![PluginsConfig::Sample(SampleConfig {
            item: "value1".into(),
        }),
        PluginsConfig::ResourceStorage(RepositoryConfig::LocalFs(
            LocalFsRepoDesc {
                dir_path: "/tmp/kbs-resource".into(),
            },
        ))],
    })]
    #[case("test_data/configs/coco-as-grpc-2.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            ..Default::default()
        },
        #[cfg(feature = "coco-as-grpc")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASGrpc(
                    crate::attestation::coco::grpc::GrpcConfig {
                        as_addr: "http://as:50004".into(),
                        pool_size: crate::attestation::coco::grpc::DEFAULT_POOL_SIZE,
                    },
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: true,
            payload_request_size: DEFAULT_PAYLOAD_REQUEST_SIZE,
            worker_count: None,
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::Simple(SimpleAdminConfig {
                personas: vec![SimplePersonaConfig {
                    id: "admin1".to_string(),
                    public_key_path: "/opt/confidential-containers/trustee/admin1-pubkey.pem".into(),
                    key_type: Default::default()
                }],
            }),
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/coco-as-builtin-2.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            trusted_certs_paths: vec![],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-builtin")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASBuiltIn(
                    attestation_service::config::Config {
                        work_dir: "/opt/confidential-containers/attestation-service".into(),
                        rvps_config: RvpsConfig::BuiltIn(RvpsCrateConfig{
                            storage: ReferenceValueStorageConfig::LocalFs(local_fs::Config{
                                file_path: "/opt/confidential-containers/attestation-service/reference_values".into(),
                            }),
                            extractors: None,
                        }),
                        attestation_token_broker: EarTokenConfiguration {
                            duration_min: 5,
                            ..Default::default()
                        },
                        verifier_config: None,
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: true,
            payload_request_size: DEFAULT_PAYLOAD_REQUEST_SIZE,
            worker_count: None,
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::InsecureAllowAll,
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/intel-ta-2.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            trusted_jwk_sets: vec!["https://portal.trustauthority.intel.com".into()],
            insecure_key: false,
            trusted_certs_paths: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "intel-trust-authority-as")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::IntelTA(
                    crate::attestation::intel_trust_authority::IntelTrustAuthorityConfig {
                        base_url: "https://api.trustauthority.intel.com".into(),
                        api_key: "tBfd5kKX2x9ahbodKV1...".into(),
                        certs_file: "https://portal.trustauthority.intel.com".into(),
                        allow_unmatched_policy: None,
                        policy_ids: vec![],
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            sockets: vec!["0.0.0.0:8080".parse().unwrap()],
            private_key: None,
            certificate: None,
            insecure_http: true,
            payload_request_size: DEFAULT_PAYLOAD_REQUEST_SIZE,
            worker_count: None,
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::DenyAll,
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/coco-as-grpc-3.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            ..Default::default()
        },
        #[cfg(feature = "coco-as-grpc")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASGrpc(
                    crate::attestation::coco::grpc::GrpcConfig {
                        as_addr: "http://127.0.0.1:50004".into(),
                        pool_size: 100,
                    },
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            insecure_http: true,
            ..Default::default()
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::DenyAll,
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/intel-ta-3.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            trusted_jwk_sets: vec!["https://portal.trustauthority.intel.com".into()],
            insecure_key: false,
            trusted_certs_paths: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "intel-trust-authority-as")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::IntelTA(
                    crate::attestation::intel_trust_authority::IntelTrustAuthorityConfig {
                        base_url: "https://api.trustauthority.intel.com".into(),
                        api_key: "tBfd5kKX2x9ahbodKV1...".into(),
                        certs_file: "https://portal.trustauthority.intel.com".into(),
                        allow_unmatched_policy: None,
                        policy_ids: vec![],
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            insecure_http: true,
            ..Default::default()
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::DenyAll,
        },
        policy_engine: PolicyEngineConfig::default(),
        plugins: Vec::new(),
    })]
    #[case("test_data/configs/coco-as-builtin-3.toml",         KbsConfig {
        attestation_token: AttestationTokenVerifierConfig {
            trusted_certs_paths: vec![],
            insecure_key: false,
            trusted_jwk_sets: vec![],
            extra_teekey_paths: vec![],
        },
        #[cfg(feature = "coco-as-builtin")]
        attestation_service: crate::attestation::config::AttestationConfig {
            attestation_service:
                crate::attestation::config::AttestationServiceConfig::CoCoASBuiltIn(
                    attestation_service::config::Config {
                        work_dir: "/opt/confidential-containers/attestation-service".into(),
                        rvps_config: RvpsConfig::BuiltIn(RvpsCrateConfig::default()),
                        attestation_token_broker: EarTokenConfiguration {
                            duration_min: 5,
                            policy_dir: "/opt/confidential-containers/attestation-service/ear-policies".into(),
                            ..Default::default()
                        },
                        verifier_config: None,
                    }
                ),
            timeout: crate::attestation::config::DEFAULT_TIMEOUT,
        },
        http_server: HttpServerConfig {
            insecure_http: true,
            ..Default::default()
        },
        admin: AdminConfig {
            admin_backend: AdminBackendType::Simple(SimpleAdminConfig {
                personas: Vec::new(),
            }),
        },
        policy_engine: PolicyEngineConfig {
            policy_path: "/opa/confidential-containers/kbs/policy.rego".into(),
        },
        plugins: vec![
        PluginsConfig::ResourceStorage(RepositoryConfig::LocalFs(
            LocalFsRepoDesc {
                dir_path: "/opt/confidential-containers/kbs/repository".into(),
            },
        ))],
    })]
    fn read_config(#[case] config_path: &str, #[case] expected: KbsConfig) {
        let config = KbsConfig::try_from(Path::new(config_path)).unwrap();
        assert_eq!(config, expected, "case {config_path}");
    }
}
